### PR TITLE
Add directory object

### PR DESCRIPTION
**Pull request by jsha**
This allows clients to be configured with a single URL that will give them information about the other URLs available for various resources.

### CODE REVIEW

**Comment by ekr** [comment]
Do we really have enough endpoints that we need this?

**Comment by jsha** [comment]
This is part of Martin's goal of achieve no coupling between endpoints: I.e. any of the "entry" endpoints can be on any URL. Richard specifically requested it. The alternatives are:

 - Enforce a path structure (i.e. all URLs are relative to some root).
 - Configure each endpoint into each client.

I prefer the former, but Martin dislikes it, so this is a fine compromise.

**Comment by ekr** [comment]
Why not just have this request provide the entire manifest of endpoints?

**Comment by bifurcation** [comment]
To be clear: I requested this because the architecture that @jsha insists on for revocation requires a "revokeCertificate" URL that is not discoverable through other processes.  If we were to allow revocation via a POST to the certificate URL, this might eventually be necessary once we had more endpoints.  But it would not be required right now, because everything is currently discoverable, transitively, starting from the new-registration endpoint.

File: draft-barnes-acme.md [hunk 1]
```diff
@@ -245,12 +245,13 @@ ACME is structured as a REST application with a few types of resources:
...
 contact information, is stored in registration resources.
 
-In order to accomplish ACME transactions, a client needs to have the server's new-registration, new-authorization, and new-certificate URIs; the remaining URIs are provided to the client as a result of requests to these URIs.  To simplify
-configuration, ACME uses the "next" link relation to indicate URI to contact for the next step in processing: From registration to authorization, and from authorization to certificate issuance.  In this way, a client need only be configured with the registration URI.
+In order to accomplish ACME transactions, a client needs to have the server's new-registration, new-authorization, new-certificate, and revoke-certificate URIs; the remaining URIs are provided to the client as a result of requests to these URIs.  To simplify
+configuration, ACME provides a directory object. The client can fetch the directory object at startup to learn the URL for each resource.
```
Comment (bifurcation) [inline draft-barnes-acme.md]: Please just rewrite this paragraph to say that ACME uses different URIs for different management functions; these functions are listed in a directory; and so clients only need to be configured with the directory URI.

File: draft-barnes-acme.md [hunk 2]
```diff
@@ -307,6 +308,31 @@ Authorization and challenge objects can also contain error information to indica
 
 TODO: Flesh out errors and syntax for them
 
+## Directory
+
+In order to help clients configure themselves with the right URLs for each ACME
+operation, ACME servers provide a directory object. This should be the root URL
+with which clients are configured. It has this structure:
+
+newRegistration (required, string): URL to the new-registration resource.
```
Comment (bifurcation) [inline draft-barnes-acme.md]: I realize we use this for other ACME objects, but this notation is a bad fit for this purpose.  Better to just
* Say that the directory is a JSON dictionary, whose keys are from a standard list of functions, and whose values are URIs for those functions (as strings)
* Define the list of functions, with cross references to the relevant sections

File: draft-barnes-acme.md [hunk 3]
```diff
@@ -307,6 +308,31 @@ Authorization and challenge objects can also contain error information to indica
...
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "newRegistration": "https://example.com/acme/new-reg",
+  "newAuthorization": "https://example.com/acme/new-authz",
+  "newRegistration": "https://example.com/acme/new-cert",
+  "newRegistration": "https://example.com/acme/revoke-cert"
+}
+
```
Comment (bifurcation) [inline draft-barnes-acme.md]: You are missing a closing "~~~~~~~~~~"

**Comment by bifurcation** [comment]
For the record, it makes me really sad that we need this so early in the game.  I would really rather not go to the effort of specifying this and implementing it in the version of this spec that Let's Encrypt uses for their initial deployment.  I honestly think there's a fair chance we could get away without needing it at all, ever.  It would be much cleaner to allow revocation by POST to the certificate URI, removing the need for this PR.

**Comment by jsha** [comment]
Updated with those changes, thanks!

File: draft-barnes-acme.md [hunk 4]
```diff
@@ -395,6 +397,30 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+~~~~~~~~~~
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "newRegistration": "https://example.com/acme/new-reg",
+  "newAuthorization": "https://example.com/acme/new-authz",
+  "newRegistration": "https://example.com/acme/new-cert",
+  "newRegistration": "https://example.com/acme/revoke-cert"
```
Comment (martinthomson) [inline draft-barnes-acme.md]: Duplicate newRegistration label.

**Comment by martinthomson** [comment]
As I described in my email, I don't like this as a solution for revocation.  I'd rather have the revocation link discoverable via the certificate (renewal) resource.

I think that registration resources are all you need.  The response to a registration can provide endpoints for requesting new authorizations of certificates.  That allows the server to bake a lot more context into paths if they choose to (whether that is at all a benefit will depend on server architecture).

File: draft-barnes-acme.md (hunk 4 above)
Comment (jsha) [inline draft-barnes-acme.md]: Thanks, fixed.

**Comment by jsha** [comment]
I think the idea that newRegistration leads to newAuthorization, etc., only really makes sense in the simplest straight-through case where you register once, authorize one domain, issue a cert for it, and then never use that registration again.

If we expect clients to get multiple authorizations, across long spans of time, and issue multiple certificates, it becomes onerous for the client to keep track of all the URLs it needs across time.

Here's a concrete example: We are planning to do versioning but putting each version on a new hostname. If a user registers with client version 1.0, against a server running 1.0, they receive a "next" link pointing at, say, `https://v1.api.letsencrypt.org/acme/new-authz`. If the new-registration link is the only way to obtain that URL, the client has to remember that URL for all future authorizations. Later, we release protocol version 2.0, and host it at `https://v2.api.letsencrypt.org/acme/new-authz`. The user upgrades their client to version 2.0, and expects to continue to use their current registration, which has all their current certs attached to it. But their new-authz URL is stale. The simplest way to refresh it is to have the client fetch a directory each time it runs.

**Comment by martinthomson** [comment]
That's hardly onerous.  You have to remember a private key to use a registration, so remembering where the registration isn't anything significant extra to remember.

Beyond that, it's optimization.  If the registration resource links to where authorizations are created, you choose whether to lookup that link or remember it.

**Comment by kuba** [comment]
It should be possible to revoke with just certificate and the corresponding key (c.f. https://github.com/letsencrypt/acme-spec/issues/85). It is far more likely to lose an account key, copies of resources, next URIs, etc., than lose the certificate and corresponding private key. "Revocation link discoverable via the certificate (renewal) resource" would not be possible if you don't have the Certificate Resource URI...

**Comment by martinthomson** [comment]
I'd like to see the new cert and new authorization resources hanging off the registration resource, not a central directory.

File: draft-barnes-acme.md [hunk 5]
```diff
@@ -262,12 +262,14 @@ ACME is structured as a REST application with a few types of resources:
 * A "new-registration" resource
 * A "new-authorization" resource
 * A "new-certificate" resource
+* A "revoke-certificate" resource
```
Comment (bifurcation) [inline draft-barnes-acme.md]: Shouldn't you add the directory here?

File: draft-barnes-acme.md [hunk 6]
```diff
@@ -395,6 +397,30 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+
+~~~~~~~~~~
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "newRegistration": "https://example.com/acme/new-reg",
+  "newAuthorization": "https://example.com/acme/new-authz",
+  "newCertificate": "https://example.com/acme/new-cert",
+  "revokeCertificate": "https://example.com/acme/revoke-cert"
```
Comment (martinthomson) [inline draft-barnes-acme.md]: These names need to match up with the names that are used in #156

Comment (bifurcation) [inline draft-barnes-acme.md]: That would also remove the need to list these things yet again here.  So +1

**Comment by jcjones** [comment]
I am happy with how this turned out. LGTM, r=jcjones

**Pull request merged**